resen/Resen.py

```diff
@@ -662,13 +662,56 @@ def create_container(self,**input_kwargs):
         if image_id not in local_image_ids:
             print("Pulling image: %s" % image_name)
             print("   This may take some time...")
-            self.docker.images.pull(pull_image)
+            DockerHelper.stream_pull_image(pull_image)
+            image = self.docker.images.get(pull_image)
+            repo,digest = pull_image.split('@')
+            # When pulling from repodigest sha256 no tag is assigned. So:
+            image.tag(repo, tag=image_name)
             print("Done!")
 
         container_id = self.docker.containers.create(image_id,**create_kwargs)
 
         return container_id.id
 
+    @staticmethod
+    def stream_pull_image(pull_image):
+        import datetime
+        def truncate_secs(delta_time, fmt=":%.2d"):
+            delta_str = str(delta_time).split(':')
+            return ":".join(delta_str[:-1]) + fmt%(float(delta_str[-1]))
+        def update_bar(sum_total,accumulated,t0,current_time, scale=0.5):
+            percentage = accumulated/sum_total*100
+            nchars = int(percentage*scale)
+            bar = "\r["+nchars*"="+">"+(int(100*scale)-nchars)*" "+"]"
+            time_info = "Elapsed time: %s"%truncate_secs(current_time - t0)
+            print(bar+" %5.2f %%, %5.3f/%4.2fGB %s"%(percentage,
+                accumulated/1024**3,sum_total/1024**3,time_info),end="")
+
+        client = docker.APIClient(base_url='unix://var/run/docker.sock')
+        id_list = []
+        id_current = []
+        id_total = 0
+        t0 = prev_time = datetime.datetime.now()
+        try:
+            for line in client.pull(pull_image,stream=True, decode=True):
+                if 'progress' not in line:
+                    continue
+                line_current = line['progressDetail']['current']
+                if line['id'] not in id_list:
+                    id_list += [line['id']]
+                    id_current += [line_current]
+                    id_total += line['progressDetail']['total']
+                else:
+                    id_current[id_list.index(line['id'])] = line_current
+                current_time = datetime.datetime.now()
+                if (current_time-prev_time).total_seconds()<1:
+                    continue
+                prev_time = current_time
+                update_bar(id_total,sum(id_current),t0,current_time)
+            update_bar(id_total,sum(id_current),t0,current_time)
+        except:
+            print("\nError pulling image %s"%pull_image)
+        print() # to avoid erasing the progress bar at the end
```

Comment:
Since we are in here adding code anyways, what happens if the image pull fails?

Can we have `stream_pull_image` return `True` or `False` depending on if it succeeds or not?

Reply:
Sure. I'll do that.